cmd/config.go

```diff
@@ -0,0 +1,109 @@
+var config Config
+
+// defaults is a Config set to default values used by Cobra
+var defaults = DefaultConfig()
+
+const defaultOtlpEndpoint = "localhost:4317"
+const spanBgSockfilename = "otel-cli-background.sock"
+
+// DefaultConfig returns a Config with all defaults set.
+func DefaultConfig() Config {
+	return Config{
+		Endpoint:               "",
+		Timeout:                "1s",
+		Headers:                map[string]string{},
+		Insecure:               false,
+		Blocking:               false,
+		NoTlsVerify:            false,
+		ServiceName:            "otel-cli",
+		SpanName:               "todo-generate-default-span-names",
+		Kind:                   "client",
+		Attributes:             map[string]string{},
+		TraceparentCarrierFile: "",
+		TraceparentIgnoreEnv:   false,
+		TraceparentPrint:       false,
+		TraceparentPrintExport: false,
+		TraceparentRequired:    false,
+		BackgroundParentPollMs: 10,
+		BackgroundSockdir:      "",
+		BackgroundWait:         false,
+		SpanStartTime:          "now",
+		SpanEndTime:            "now",
+		EventName:              "todo-generate-default-event-names",
+		EventTime:              "now",
+		CfgFile:                "",
+		Verbose:                false,
+	}
+}
+
+// Config stores the runtime configuration for otel-cli.
+// This is used as a singleton as "config" and accessed from many other files.
+// Data structure is public so that it can serialize to json easily.
+type Config struct {
+	Endpoint    string            `json:"endpoint"`
+	Timeout     string            `json:"timeout"`
+	Headers     map[string]string `json:"headers"` // TODO: needs json marshaler hook to mask tokens
+	Insecure    bool              `json:"insecure"`
+	Blocking    bool              `json:"blocking"`
+	NoTlsVerify bool              `json:"no_tls_verify"`
+
+	ServiceName string            `json:"service_name"`
+	SpanName    string            `json:"span_name"`
+	Kind        string            `json:"span_kind"`
+	Attributes  map[string]string `json:"span_attributes"`
+
+	TraceparentCarrierFile string `json:"traceparent_carrier_file"`
+	TraceparentIgnoreEnv   bool   `json:"traceparent_ignore_env"`
+	TraceparentPrint       bool   `json:"traceparent_print"`
+	TraceparentPrintExport bool   `json:"traceparent_print_export"`
+	TraceparentRequired    bool   `json:"traceparent_required"`
+
+	BackgroundParentPollMs int    `json:"background_parent_poll_ms"`
+	BackgroundSockdir      string `json:"background_socket_directory"`
+	BackgroundWait         bool   `json:"background_wait"`
+
+	SpanStartTime string `json:"span_start_time"`
+	SpanEndTime   string `json:"span_end_time"`
+	EventName     string `json:"event_name"`
+	EventTime     string `json:"event_time"`
+
+	CfgFile string `json:"config_file"`
+	Verbose bool   `json:"verbose"`
+}
+
+// ToStringMap flattens the configuration into a stringmap that is easy to work
+// with in tests especially with cmp.Diff. See test_main.go.
+func (c Config) ToStringMap() map[string]string {
```

Comment:
this is good because it actually maps to how viper stores keys in its env dir anyways, so props to utility here